### PR TITLE
machine.memX raises exception stating slices aren't supported instead of crashing

### PR DESCRIPTION
**Pull request by Arrowana**
fix for #6588

### CODE REVIEW

File: extmod/machine_mem.c [hunk 1]
```diff
@@ -63,6 +63,9 @@ STATIC void machine_mem_print(const mp_print_t *print, mp_obj_t self_in, mp_prin
 STATIC mp_obj_t machine_mem_subscr(mp_obj_t self_in, mp_obj_t index, mp_obj_t value) {
     // TODO support slice index to read/write multiple values at once
     machine_mem_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    if (mp_obj_is_type(index, &mp_type_tuple)) {
+        mp_raise_TypeError(MP_ERROR_TEXT("Does not support slices"));
```
Comment (dlech) [inline extmod/machine_mem.c:67]: How about raising `NotImplementedError` instead since it seems like the intention is to implement it eventually?

Comment (Arrowana) [reply extmod/machine_mem.c]: Good point, fixed that.
I also added a test for it

**Comment by dpgeorge** [comment]
Thanks for the contribution.

Actually, it also currently crashes when any non-int object is passed in, eg `machine.mem8['foo']`.  And using bool should technically work (in Python False/True are equivalent to 0/1 when ints are accepted) but this also crashes: `machine.mem8[False]`.

Probably these should all be fixed because, as you mention, it's hard to debug hard crashes like this.  A very simple fix would be to replace the use of `mp_obj_int_get_truncated()` with `mp_obj_get_int_truncated()` which will do a proper type check on the argument, and also convert bool to int.

There would be a (very) small performance overhead using `mp_obj_get_int_truncated()` (compared to using `mp_obj_int_get_truncated()`) but the only way to avoid that is to just leave it as-as and allow it to crash if a non-int is passed it.  Well, since the `machine.memX` objects are more for convenience than performance (there are many other ways to read/write memory in a faster way), I think it's OK to have the check and use `mp_obj_int_get_truncated()`.

**Comment by Arrowana** [comment]
@dpgeorge That is a much better approach that does not have edge cases. I implemented as advised

File: extmod/machine_mem.c [hunk 2]
```diff
@@ -85,7 +85,7 @@ STATIC mp_obj_t machine_mem_subscr(mp_obj_t self_in, mp_obj_t index, mp_obj_t va
     } else {
         // store
         uintptr_t addr = MICROPY_MACHINE_MEM_GET_WRITE_ADDR(index, self->elem_size);
-        uint32_t val = mp_obj_get_int_truncated(value);
+        uint32_t val = mp_obj_int_get_truncated(value);
```
Comment (dpgeorge) [inline extmod/machine_mem.c:88]: why is this change made? this will now make something like `machine.mem32[0] = "123"` crash

**Comment by dpgeorge** [comment]
Thanks for updating, squashed and merged in 922f81dfd1b94e23f23f05aac0ecafac8645ce46

**Pull request closed without merging**